File: Data Structures & Algorithms/top-k-elements-in-list/submission-6.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        
        // use hashmap to create count 
        HashMap<Integer, Integer> map = new HashMap<>();
        for(int num: nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }

        // use minHeap to store values 
        PriorityQueue<Integer> minHeap = new PriorityQueue<>((a, b) -> 
            map.get(a) - map.get(b)
        );

        for (Map.Entry<Integer, Integer> entry: map.entrySet()){
            minHeap.add(entry.getKey());

            if (minHeap.size() > k) {
                minHeap.poll();
            }
        }


        int[] res = new int[k];
        for(int i = 0; i < k; i++) {
            res[i] = minHeap.poll();
        }

        return res;
    }
}

File: Data Structures & Algorithms/permutations/submission-6.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        
        self.res = []
        self.backtrack(nums, set(), [])
        return self.res 
    

    def backtrack(self, nums, seen, so_far):

        if len(so_far) == len(nums):
            self.res.append(list(so_far))
            return 

        for num in nums:
            if num not in seen:
                so_far.append(num)
                seen.add(num)

                self.backtrack(nums, seen, so_far)

                so_far.pop() 
                seen.remove(num)

